casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala

```diff
@@ -37,72 +37,52 @@ object InterpreterUtil {
       dag: BlockDagRepresentation[F],
       runtimeManager: RuntimeManager[F]
   ): F[BlockProcessing[Option[StateHash]]] = {
-    val preStateHash    = ProtoUtil.preStateHash(b)
-    val tsHash          = ProtoUtil.tuplespace(b)
-    val deploys         = ProtoUtil.deploys(b)
-    val internalDeploys = deploys.map(InternalProcessedDeploy.fromProcessedDeploy)
-    val timestamp       = b.header.timestamp
-    val blockNumber     = b.body.state.blockNumber
+    val incomingPreStateHash = ProtoUtil.preStateHash(b)
+    val tsHash               = ProtoUtil.tuplespace(b)
+    val deploys              = ProtoUtil.deploys(b)
+    val internalDeploys      = deploys.map(InternalProcessedDeploy.fromProcessedDeploy)
+    val timestamp            = b.header.timestamp
+    val blockNumber          = b.body.state.blockNumber
     for {
-      _                    <- Span[F].mark("before-unsafe-get-parents")
-      parents              <- ProtoUtil.getParents[F](b)
-      _                    <- Span[F].mark("before-compute-parents-post-state")
-      possiblePreStateHash <- computeParentsPostState[F](parents, dag, runtimeManager)
-      _                    <- Log[F].info(s"Computed parents post state for ${PrettyPrinter.buildString(b)}.")
-      invalidBlocksSet     <- dag.invalidBlocks
-      unseenBlocksSet      <- ProtoUtil.unseenBlockHashes(dag, b)
+      _                  <- Span[F].mark("before-unsafe-get-parents")
+      parents            <- ProtoUtil.getParents(b)
+      _                  <- Span[F].mark("before-compute-parents-post-state")
+      preStateHashEither <- computeParentsPostState(parents, dag, runtimeManager).attempt
+      _                  <- Log[F].info(s"Computed parents post state for ${PrettyPrinter.buildString(b)}.")
+      invalidBlocksSet   <- dag.invalidBlocks
+      unseenBlocksSet    <- ProtoUtil.unseenBlockHashes(dag, b)
       seenInvalidBlocksSet = invalidBlocksSet.filterNot(
         block => unseenBlocksSet.contains(block.blockHash)
       ) // TODO: Write test in which switching this to .filter makes it fail
       invalidBlocks = seenInvalidBlocksSet.map(block => (block.blockHash, block.sender)).toMap
       _             <- Span[F].mark("before-process-pre-state-hash")
-      result <- processPossiblePreStateHash[F](
-                 runtimeManager,
-                 preStateHash,
-                 tsHash,
-                 internalDeploys,
-                 possiblePreStateHash,
-                 BlockData(timestamp, blockNumber),
-                 invalidBlocks,
-                 isGenesis = b.header.parentsHashList.isEmpty
-               )
+      isGenesis     = b.header.parentsHashList.isEmpty
```

Comment:
I am not sure if I'm reading it right - but my sense is that we're doing unnecessary work if `preStateHashEither` is `Left` here

Reply:
So looks like the whole thing might be rewritten as (I just rearranged it, perhaps it can be written in a nicer way)

```scala
    for {
      _                  <- Span[F].mark("before-unsafe-get-parents")
      parents            <- ProtoUtil.getParents(b)
      _                  <- Span[F].mark("before-compute-parents-post-state")
      preStateHashEither <- computeParentsPostState(parents, dag, runtimeManager).attempt
      _                  <- Log[F].info(s"Computed parents post state for ${PrettyPrinter.buildString(b)}.")
      result <- preStateHashEither match {
                 case Left(ex) =>
                   BlockStatus.exception(ex).asLeft[Option[StateHash]].pure
                 case Right(computedPreStateHash)
                     if (incomingPreStateHash == computedPreStateHash) =>
                   for {
                     invalidBlocksSet <- dag.invalidBlocks
                     unseenBlocksSet  <- ProtoUtil.unseenBlockHashes(dag, b)
                     seenInvalidBlocksSet = invalidBlocksSet.filterNot(
                       block => unseenBlocksSet.contains(block.blockHash)
                     ) // TODO: Write test in which switching this to .filter makes it fail
                     invalidBlocks = seenInvalidBlocksSet
                       .map(block => (block.blockHash, block.sender))
                       .toMap
                     _         <- Span[F].mark("before-process-pre-state-hash")
                     isGenesis = b.header.parentsHashList.isEmpty
                     result <- replayBlockDeploys(
                                runtimeManager,
                                incomingPreStateHash,
                                tsHash,
                                internalDeploys,
                                BlockData(timestamp, blockNumber),
                                invalidBlocks,
                                isGenesis
                              )
                   } yield result
                 case Right(computedPreStateHash) => //TODO at this point we may just as well terminate the replay, there's no way it will succeed.
                   Log[F].warn(
                     s"Computed pre-state hash ${PrettyPrinter.buildString(computedPreStateHash)} does not equal block's pre-state hash ${PrettyPrinter
                       .buildString(incomingPreStateHash)}"
                   ) >> none[StateHash].asRight[BlockError].pure
               }
    } yield result
```

Reply:
> we're doing unnecessary work if preStateHashEither is Left here

Indeed we are. Since I find nested for-blocks ugly, I went ahead and extracted a method for the nested one (or rather: pushed that logic into the existing `replayBlockDeploys`). Thanks for the nudge, I think it's been fruitful! :)

Process remark: normally, when some deficiency was already there before the PR, and the PR merely moves it (does not introduce it) - we'd require a fix for the deficiency only if its critical and approve the PR if it's good otherwise. Even when it's critical, we'd sometimes consider a separate follow-up PR for fixing it.